apps/shell/app.js

```diff
@@ -51,6 +68,79 @@ app.set('views', path.join(__dirname, 'views'));
 // Mount the routes at the base URI
 app.use(process.env.PASSENGER_BASE_URI || '/', router);
 
+var terminals = {
+
+  instances: {
+
+  },
+
+  sessionsInfo: function () {
+
+    return Object.entries(this.instances)
+          .sort()
+          .map(function(array){ return {id: array[0], host: array[1].host }; });
+  },
+
+  create: function (host, dir, uuid) {
+    var cmd = 'ssh';
+    var args = dir ? [host, '-t', 'cd \'' + dir.replace(/\'/g, "'\\''") + '\' ; exec ${SHELL} -l'] : [host];
+
+    process.env.LANG = 'en_US.UTF-8'; // this patch (from b996d36) lost when removing wetty (2c8a022)
+
+    this.instances[uuid] = {term: pty.spawn(cmd, args, {
+      name: 'xterm-256color',
+      cols: 80,
+      rows: 30
+    }), host: host}
+
+    return uuid;
+  },
+
+  exists: function (uuid) {
+    if (uuid in this.instances) {
+      return true;
+    } else {
+      return false;
+    }
+  },
+
+  get: function (uuid) {
+    return this.instances[uuid].term;
+  },
+
+  attach: function (uuid, ws) {
+    var term = this.get(uuid);
+    term.resume();
+    term.resize(80, 30);
+
+    term.on('data', function (data) {
+      ws.send(data, function (error) {
+        if (error) console.log('Send error: ' + error.message);
+      });
+    });
+
+    term.on('error', function (error) {
+      ws.close();
+    });
+
+    term.on('close', function () {
+      ws.close();
+    });
+
+    ws.on('message', function (msg) {
+      msg = JSON.parse(msg);
+      if (msg.input)  term.write(msg.input);
+      if (msg.resize) term.resize(parseInt(msg.resize.cols), parseInt(msg.resize.rows));
+    });
+
+    ws.on('close', function () {
+      term.pause();
+      console.log('Closed terminal: ' + term.pid);
```

Comment:
"Closed terminal" should be "Pause terminal"

Reply:
I believe this should have access to a CloseEvent. So `ws.on('close', function(event) {`. And I think this event should be similar to https://developer.mozilla.org/en-US/docs/Web/API/CloseEvent. Which means, possibly that if `event.code == 1000` or `1001` that is the user manually closing, but if `event.code != 1000` or `1001` then we have the error cases we want to protect against. In that case, what we could do is do `term.pause()` it is not a "Normal Closure" or "Going Away", but do the preexisting `term.close()` if it is.